File: clients/sellingpartner-api-aa-java/src/com/amazon/SellingPartnerAPIAA/LWAClient.java
package com.amazon.SellingPartnerAPIAA;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.squareup.okhttp.MediaType;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.RequestBody;
import com.squareup.okhttp.Response;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.EnumUtils;

class LWAClient {
    private static final String ACCESS_TOKEN_KEY = "access_token";
    private static final String ACCESS_TOKEN_EXPIRES_IN = "expires_in";
    private static final MediaType JSON_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");
    @Getter
    private String endpoint;

    @Setter(AccessLevel.PACKAGE)
    private OkHttpClient okHttpClient;
    private LWAAccessTokenCache lwaAccessTokenCache;

    /**
     * Sets cache to store access token until token is expired
     */
    public void setLWAAccessTokenCache(LWAAccessTokenCache tokenCache) {
        this.lwaAccessTokenCache = tokenCache;
    }

    LWAClient(String endpoint) {
        okHttpClient = new OkHttpClient();
        this.endpoint = endpoint;
    }

    String getAccessToken(LWAAccessTokenRequestMeta lwaAccessTokenRequestMeta) throws LWAException {
        if (lwaAccessTokenCache != null) {
            return getAccessTokenFromCache(lwaAccessTokenRequestMeta);
        } else {
            return getAccessTokenFromEndpoint(lwaAccessTokenRequestMeta);
        }
    }

    String getAccessTokenFromCache(LWAAccessTokenRequestMeta lwaAccessTokenRequestMeta) throws LWAException {
        String accessTokenCacheData = (String) lwaAccessTokenCache.get(lwaAccessTokenRequestMeta);
        if (accessTokenCacheData != null) {
            return accessTokenCacheData;
        } else {
            return getAccessTokenFromEndpoint(lwaAccessTokenRequestMeta);
        }
    }

    String getAccessTokenFromEndpoint(LWAAccessTokenRequestMeta lwaAccessTokenRequestMeta) throws LWAException {
        RequestBody requestBody = RequestBody.create(JSON_MEDIA_TYPE, new Gson().toJson(lwaAccessTokenRequestMeta));
        Request accessTokenRequest = new Request.Builder().url(endpoint).post(requestBody).build();
        LWAExceptionErrorCode lwaErrorCode = null;
        String accessToken;
        try {
            Response response = okHttpClient.newCall(accessTokenRequest).execute();
            JsonObject responseJson = new JsonParser().parse(response.body().string()).getAsJsonObject();
            if (!response.isSuccessful()) {
                // Check if response has element error and is a known LWA error code
                if (responseJson.has("error") &&
                        EnumUtils.isValidEnum(LWAExceptionErrorCode.class, responseJson.get("error").getAsString())) {
                    throw new LWAException(responseJson.get("error").getAsString(),
                            responseJson.get("error_description").getAsString(), "Error getting LWA Token");
                } else { // else throw other LWA error
                    throw new LWAException(LWAExceptionErrorCode.other.toString(), "Other LWA Exception",
                            "Error getting LWA Token");
                }
            }
            accessToken = responseJson.get(ACCESS_TOKEN_KEY).getAsString();
            if (lwaAccessTokenCache != null) {
                long timeToTokenexpiry = responseJson.get(ACCESS_TOKEN_EXPIRES_IN).getAsLong();
                lwaAccessTokenCache.put(lwaAccessTokenRequestMeta, accessToken, timeToTokenexpiry);
            }
        } catch (LWAException e) { // throw LWA exception
            throw new LWAException(e.getErrorCode(), e.getErrorMessage(), e.getMessage());
        } catch (Exception e) { // throw other runtime exceptions
            throw new RuntimeException("Error getting LWA Token");
        }
        return accessToken;
    }
}

File: clients/sellingpartner-api-aa-php/src/authandauth/LWAClient.php
<?php

namespace SpApi\AuthAndAuth;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Exception\BadResponseException;
use GuzzleHttp\Exception\GuzzleException;
use RuntimeException;
use Exception;

class LWAClient
{
    private Client $client;

    private string $endpoint;

    private ?LWAAccessTokenCache $lwaAccessTokenCache = null;

    public function __construct(string $endpoint)
    {
        $this->client = new Client();
        $this->endpoint = $endpoint;
    }

    public function setLWAAccessTokenCache(?LWAAccessTokenCache $tokenCache): void
    {
        $this->lwaAccessTokenCache = $tokenCache;
    }

    public function getAccessToken(LWAAccessTokenRequestMeta &$lwaAccessTokenRequestMeta): string
    {
        if ($this->lwaAccessTokenCache !== null) {
            return $this->getAccessTokenFromCache($lwaAccessTokenRequestMeta);
        } else {
            return $this->getAccessTokenFromEndpoint($lwaAccessTokenRequestMeta);
        }
    }

    public function getAccessTokenFromCache(LWAAccessTokenRequestMeta &$lwaAccessTokenRequestMeta)
    {
        $requestBody = json_encode($lwaAccessTokenRequestMeta);
        if (!$requestBody) {
            throw new RuntimeException("Request body could not be encoded");
        }
        $accessTokenCacheData = $this->lwaAccessTokenCache->get($requestBody);
        if ($accessTokenCacheData !== null) {
            return $accessTokenCacheData;
        } else {
            return $this->getAccessTokenFromEndpoint($lwaAccessTokenRequestMeta);
        }
    }

    public function getAccessTokenFromEndpoint(LWAAccessTokenRequestMeta &$lwaAccessTokenRequestMeta)
    {
        $requestBody = json_encode($lwaAccessTokenRequestMeta);

        if (!$requestBody) {
            throw new RuntimeException("Request body could not be encoded");
        }

        $contentHeader = [
            "Content-Type" => "application/json",
        ];

        try {
            $lwaRequest = new Request("POST", $this->endpoint, $contentHeader, $requestBody);

            $lwaResponse = $this->client->send($lwaRequest);
            $responseJson = json_decode($lwaResponse->getBody(), true);

            if (!$responseJson["access_token"] || !$responseJson["expires_in"]) {
                throw new RuntimeException("Response did not have required body");
            }

            $accessToken = $responseJson["access_token"];

            if ($this->lwaAccessTokenCache !== null) {
                $timeToTokenExpire = (float)$responseJson["expires_in"];
                $this->lwaAccessTokenCache->set($requestBody, $accessToken, $timeToTokenExpire);
            }
        } catch (BadResponseException $e) {
            //Catches 400 and 500 level error codes
            throw new RuntimeException("Unsuccessful LWA token exchange", $e->getCode());
        } catch (Exception $e) {
            throw new RuntimeException("Error getting LWA Access Token", $e->getCode());
        } catch (GuzzleException $e) {
            throw new RuntimeException("Error getting LWA Access Token", $e->getCode());
        }

        return $accessToken;
    }

    public function setClient(Client $client): void
    {
        $this->client = $client;
    }

    public function getEndpoint(): string
    {
        return $this->endpoint;
    }
}
